### ISSUE TITLE
[array-type] False negative when using 'array-simple'

### ISSUE THREAD

**Issue by Shinigami92** [typescript-eslint/typescript-eslint#2323]
**Repro**

```JSON
{
  "rules": {
    "@typescript-eslint/array-type": ["warn", { "default": "array-simple", "readonly": "generic" }]
  }
}
```

```TS
public get selectableTemplates(): (ExpandedTextTemplateModel & IsSelectable)[] {
  return this.templates.filter(({ isSelectable }) => isSelectable);
}
```

**Expected Result**

Show warning and suggest to write `Array<ExpandedTextTemplateModel & IsSelectable>` instead

**Actual Result**

No warning shown

**Additional Info**

This is working as expected and logs many (~116) warnings

```JSON
{
  "rules": {
    "@typescript-eslint/array-type": ["warn"]
  }
}
```

So I think the problem/bug has something to do with how I setup my preferred option object 🤔 

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `3.7.0` |
| `@typescript-eslint/parser`        | `3.7.0` |
| `TypeScript`                       | `3.9.7` |
| `ESLint`                           | `7.5.0` |
| `node`                             | `v14.1.0` |
| `npm`                              | `6.14.4` |

**Comment by Shinigami92** [comment]
When I just remove `, "readonly": "generic"` I see warnings again 🤔

**Comment by Shinigami92** [comment]
Also it seems that _my_ case has no test defined ^^'
https://github.com/typescript-eslint/typescript-eslint/blob/fd90e31019afb81aee8f83d9632096236069520b/packages/eslint-plugin/tests/rules/array-type.test.ts

**Comment by Shinigami92** [comment]
I'm new to the project's code and don't know how to debug or test only this rule.
So it's hard to test, cause I always running **17197** tests...

I guess when using `{ default: 'array-simple', readonly: 'generic' }`, following code snippets run into returning out of the function. But I'm not sure 🤷‍♂️

`TSArrayType(node): void`
https://github.com/typescript-eslint/typescript-eslint/blob/fd90e31019afb81aee8f83d9632096236069520b/packages/eslint-plugin/src/rules/array-type.ts#L207-L212

```js
isReadonlyGeneric => true
isReadonlyArray => false

(isReadonlyGeneric && !isReadonly) => maybe true or false, depends on isReadonly
(isReadonlyArray && isReadonly) => always false
```

---

`TSTypeReference(node): void`
https://github.com/typescript-eslint/typescript-eslint/blob/fd90e31019afb81aee8f83d9632096236069520b/packages/eslint-plugin/src/rules/array-type.ts#L275-L281

```js
readonlyOption => 'generic'
defaultOption => 'array-simple'

!(isArrayType || isReadonlyArrayType) => depends on isArrayType and isReadonlyArrayType
(readonlyOption === 'generic' && isReadonlyArrayType) => depends on isReadonlyArrayType
(defaultOption === 'generic' && !isReadonlyArrayType) => always false
```

**Comment by bradzacher** [comment]
There's a [vscode launch config](https://github.com/typescript-eslint/typescript-eslint/blob/fd90e31019afb81aee8f83d9632096236069520b/.vscode/launch.json#L7-L38) which will allow you to run any given singular test in the vscode debugger.
Failing that we just use `jest`, so you can simply `cd` into the package directory and use [jest to filter to a specific test](https://jestjs.io/docs/en/cli#jest-regexfortestfiles).

We don't do any magic on top of the commands like other repos do. [`yarn test` just maps directly to `yarn jest`](https://github.com/typescript-eslint/typescript-eslint/blob/fd90e31019afb81aee8f83d9632096236069520b/packages/eslint-plugin/package.json#L41)

---

From what I can see, it looks like you're correct - the logic for the option checks looks wrong.
The options checks actually look really weird and wrong. Might be worth just revisiting them and making that logic all more sane.

**Comment by Shinigami92** [comment]
Oh my bad!
Because I'm working on Windows @home, I tried `yarn test tests\rules\array-type.test.ts`, but then no tests were found.
Now I tried `yarn test "tests/rules/array-type.test.ts"` and that works :)
But I even could simplify this to just `yarn test array-type` and this results the same ^^

Thanks for leading the way, now I can `console.log` the 💩 out of it 😅

**Comment by Shinigami92** [comment]
Ok, I have now created a matrix that lists all possible combinations and their expected results

The first two columns represent the input values
The header of column 3-6 are cases and the column list the expected outcome

| defaultOption | readonlyOption | number[]      | Array<number> | Array<Foo & Bar> | readonly number[]     | ReadonlyArray<number> | ReadonlyArray<Foo & Bar> |
| ------------- | -------------- | ------------- | ------------- | ---------------- | --------------------- | --------------------- | ------------------------ |
| array         |                | number[]      | number[]      | (Foo & Bar)[]    | readonly number[]     | readonly number[]     | readonly (Foo & Bar)[]   |
| array         | array          | number[]      | number[]      | (Foo & Bar)[]    | readonly number[]     | readonly number[]     | readonly (Foo & Bar)[]   |
| array         | array-simple   | number[]      | number[]      | (Foo & Bar)[]    | readonly number[]     | readonly number[]     | ReadonlyArray<Foo & Bar> |
| array         | generic        | number[]      | number[]      | (Foo & Bar)[]    | ReadonlyArray<number> | ReadonlyArray<number> | ReadonlyArray<Foo & Bar> |
| array-simple  |                | number[]      | number[]      | Array<Foo & Bar> | readonly number[]     | readonly number[]     | ReadonlyArray<Foo & Bar> |
| array-simple  | array          | number[]      | number[]      | Array<Foo & Bar> | readonly number[]     | readonly number[]     | readonly (Foo & Bar)[]   |
| array-simple  | array-simple   | number[]      | number[]      | Array<Foo & Bar> | readonly number[]     | readonly number[]     | ReadonlyArray<Foo & Bar> |
| array-simple  | generic        | number[]      | number[]      | Array<Foo & Bar> | ReadonlyArray<number> | ReadonlyArray<number> | ReadonlyArray<Foo & Bar> |
| generic       |                | Array<number> | Array<number> | Array<Foo & Bar> | ReadonlyArray<number> | ReadonlyArray<number> | ReadonlyArray<Foo & Bar> |
| generic       | array          | Array<number> | Array<number> | Array<Foo & Bar> | readonly number[]     | readonly number[]     | readonly (Foo & Bar)[]   |
| generic       | array-simple   | Array<number> | Array<number> | Array<Foo & Bar> | readonly number[]     | readonly number[]     | ReadonlyArray<Foo & Bar> |
| generic       | generic        | Array<number> | Array<number> | Array<Foo & Bar> | ReadonlyArray<number> | ReadonlyArray<number> | ReadonlyArray<Foo & Bar> |

Is the matrix correct or have I done something wrong?

**Comment by Shinigami92** [comment]
https://github.com/typescript-eslint/typescript-eslint/blob/fd90e31019afb81aee8f83d9632096236069520b/packages/eslint-plugin/src/rules/array-type.ts#L25

This line does always pass, also if input is something like `(string & number)[]`, but `string & number` is not "simple" 🤔

**Comment by bradzacher** [comment]
Your matrix looks correct to me. I don't think anyone's ever gone so far as to enumerate the options like that. That's a great way to conceptualise it!

It looks like the rule's logic is a bit wrong in how it handles things. Or at least it's not clear.

`(string & number)` should match `TSParenthesizedType`, which is not a "simple" type.
The rule's source looks like it handles things a bit weird. The separate `readonly` option was added in after the rule's creation, so it was probably just built wrong.

Feel free to gut and rewrite the rule however you see fit to fix this issue!

**Comment by Shinigami92** [comment]
I _worked_ on this code and researched it for about 3 hours today
And I also think that it may be better and easier to rewrite the whole rule

There is also the problem that the error messages do not always fit well, e.g. if it should be readonly it is suggested to use `Array <T>` instead of `ReadonlyArray<T>`

But such changes are a **minor** change instead of a **bug fix** 😅

---

Ok, give me a couple of days and I'll try to tackle this 💪

But I think I will also alter tests and add more tests

**Comment by bradzacher** [comment]
There's no requirement to only treat this as a bugfix.
Feel free to do whatever you think. If it improves the rule _and_ fixes the bug then that's a net win for everyone :)

Thanks for putting so much care and effort into this!

**Comment by Shinigami92** [comment]
Aside whats currently made, what do you expect `{default:'array'}` `--fix` should made?

```
// Input
let numArrArr: Array<Array<number>> = [[1], [2, 3]];



// with message "Array type using 'Array<Array<number>>' is forbidden. Use 'Array<number>[]' instead."
Array<number>[]
// no further warings 🤔 


// with message "Array type using 'Array<Array<number>>' is forbidden. Use 'Array<number[]>' instead."
Array<number[]>
// no further warings 🤔 


// with message "Array type using 'Array<Array<number>>' is forbidden. Use 'number[][]' instead."
// no second message, jump directly to number[][]
number[][]


// with message1 "Array type using 'Array<Array<number>>' is forbidden. Use 'Array<number>[]' instead."
// with message2 "Array type using 'Array<number>[]' is forbidden. Use 'number[][]' instead."
number[][]


// with message1 "Array type using 'Array<Array<number>>' is forbidden. Use 'Array<number[]>' instead."
// with message2 "Array type using 'Array<number[]>' is forbidden. Use 'number[][]' instead."
number[][]
```

<img src="https://media1.tenor.com/images/41509bd1891be615a80bf1b12eb38a0c/tenor.gif" width="200px"/>

**Comment by bradzacher** [comment]
This rule only needs to be concerned with one array type at a time.
ESLint should handle applying multiple fixers that overlap.

I.e. for `{default:'array'}`, this should happen:

```ts
  type T = number[][][];
//         ^^^^^^^^ error 1     - wants to fix to Array<number>[][]
//         ^^^^^^^^^^ error 2   - wants to fix to Array<number[]>[]
//         ^^^^^^^^^^^^ error 3 - wants to fix to Array<number[][]>
```

The important thing here is we should avoid using the `replaceText` fixer - as this creates a large fix.

Instead, for example the fixers should be:

```ts
  type T = number[][][];
//         ^^^^^^^^ error 1     - wants to insert "Array<" before number, and replace "[]" with ">"
//         ^^^^^^^^^^ error 2   - wants to insert "Array<" before number[], and replace "[]" with ">"
//         ^^^^^^^^^^^^ error 3 - wants to insert "Array<" before number[][], and replace "[]" with ">"
```

This makes it easier for ESLint to apply the fixers in order.

**Comment by Shinigami92** [comment]
So it is 5 - from inner to outer I think?

Beside that, your example confuses me a bit ^^'

`{default:'array'}` should not print any errors for `number[][][]` cause this is already the best form for `'array'`
Your errors should appear when using `'generic'`

---

But damnd this is a hard one task to tackle XD
If you want and have time, you can try in parallel to me working on this task
Cause I currently don't know if I can handle and fix this easily within a timeframe of 1-2 weeks 😅

**Comment by bradzacher** [comment]
sorry, yes - i meant `generic` not `array`!

The ESLint docs provide guidance on fixes:
https://eslint.org/docs/developer-guide/working-with-rules#applying-fixes

In a nutshell, ESLint does all of the heavy lifting of applying fixes to the file, resolving conflicts, etc.
When you ask ESLint to fix a problem via the `--fix` command, ESLint will actually do multiple passes over the file so that it can apply all fixers, as well as fix any errors that arrise as part of those fixers.

A simple example, imagine you have a fixer which adds quotes to object keys:
```
({ foo: 1 })
// fixes to
({ 'foo': 1 })
```
But what happens if the user prefers double quotes? Well then the `quotes` eslint rule fires. But because ESLint does multiple passes of fixes, eslint will apply the above fix, scan and find the `quotes` problem, apply its fix, and then scan and attempt t o find any more lint rules again.

So in our case, it's not necessary for the rule itself to care that a user might write `number[][][]` and that it should be fixed to `Array<Array<Array<number>>>`. Instead the rule just needs to care about one case at a time:
1) It needs to care that `number[]` should be fixed to `Array<number>`
1) It needs to care that `number[][]` should be fixed to `Array<number[]>`
1) It needs to care that `number[][][]` should be fixed to `Array<number[][]>`

ESLint will handle the rest.

In terms of _order_ - outside in vs inside out - that's dependent on what the AST produces. In this case the AST is constructed outside in, so that's the order in which the rule will report errors.

---

In regards to my second point about which fix API to use.
There's a few methods on the `fix` object (detailed in https://eslint.org/docs/developer-guide/working-with-rules#applying-fixes).
The thing we should do is attempt to use the `insertTextBefore` and `insertTextAfter` methods instead of the `replaceText` method.

This is so that each fix is as small as possible, to make it easier for ESLint to merge the fixes and apply it all at once.

**Comment by Shinigami92** [comment]
Just want to let you know that I may come later to this again
Currently I'm involved in other projects and next week I'm in vacation (maybe I have time there, maybe I will spend more time with IRL humans 😅)

**Comment by Shinigami92** [comment]
@bradzacher I'm so sorry
I'm to much involved in other tasks and don't find time to solve this issue

I hope someone else can benefit from my research an fix this anytime

**Comment by jsone-studios** [comment]
I going to work on this issue 🚀
